ibtk/src/utilities/SnapshotCache.cpp

```diff
@@ -0,0 +1,251 @@
+    : d_object_name(std::move(object_name))
+{
+    auto var_db = VariableDatabase<NDIM>::getDatabase();
+    d_ctx = var_db->getContext(d_object_name + "::context");
+
+    if (input_db)
+    {
+        d_snapshot_refine_type = input_db->getStringWithDefault("refine_type", d_snapshot_refine_type);
+        if (input_db->keyExists("gcw")) input_db->getIntegerArray("gcw", &d_gcw[0], NDIM);
+        d_depth = input_db->getIntegerWithDefault("depth", d_depth);
+    }
+}
+
+template <class VariableType>
+SnapshotCache<VariableType>::~SnapshotCache()
+{
+    clearSnapshots();
+}
+
+template <class VariableType>
+void
+SnapshotCache<VariableType>::clearSnapshots()
+{
+    // Clear the map.
+    d_idx_time_map.clear();
+    auto var_db = VariableDatabase<NDIM>::getDatabase();
+    // Deallocate all patch data.
+    for (const auto& idx : d_snapshot_idxs)
+    {
+        const Pointer<PatchHierarchy<NDIM> >& hierarchy = d_snapshot_hierarchies[idx];
+        int coarsest_ln = 0;
+        int finest_ln = hierarchy->getFinestLevelNumber();
+        for (int ln = coarsest_ln; ln <= finest_ln; ++ln)
+        {
+            Pointer<PatchLevel<NDIM> > level = hierarchy->getPatchLevel(ln);
+            if (level->checkAllocated(idx)) level->deallocatePatchData(idx);
+        }
+        // Delete variable from variable database
+        var_db->removePatchDataIndex(idx);
+    }
+
+    // Clear snapshots
+    d_snapshot_hierarchies.clear();
+    d_snapshot_idxs.clear();
+    d_snapshot_vars.clear();
+    d_num_snapshots_stored = 0;
+}
+
+template <class VariableType>
+void
+SnapshotCache<VariableType>::setSnapshot(const int u_idx,
+                                         const double time,
+                                         Pointer<PatchHierarchy<NDIM> > current_hierarchy)
+{
+    ++d_num_snapshots_stored;
+    // Push back a new variable
+    Pointer<VariableType> snapshot_var =
+        new VariableType(d_object_name + "::SnapshotVar_" + std::to_string(d_num_snapshots_stored), d_depth);
+    auto var_db = VariableDatabase<NDIM>::getDatabase();
+    int snapshot_idx = var_db->registerVariableAndContext(snapshot_var, d_ctx, d_gcw);
+    d_snapshot_idxs.push_back(snapshot_idx);
+    d_snapshot_vars[snapshot_idx] = snapshot_var;
+
+    // Now fill in snapshot
+    fillSnapshot(u_idx, current_hierarchy, snapshot_idx, time);
+}
+
+template <class VariableType>
+void
+SnapshotCache<VariableType>::updateSnapshot(const int u_idx,
+                                            const double time,
+                                            Pointer<PatchHierarchy<NDIM> > current_hierarchy,
+                                            const double tol)
+{
+    // Make sure we are on a stored snapshot.
+    auto it =
+        std::find_if(d_idx_time_map.begin(),
+                     d_idx_time_map.end(),
+                     [time, tol](const std::pair<int, double>& t) -> bool { return std::abs(t.second - time) < tol; });
```

Comment:
Should we have a standard "fuzzy equals" (with absolute and relative tolerance variants) that we use in these situations?

Reply:
Oh yeah. I wrote those back in #1177. We should just use that.